workspaces/optic-engine/src/projections/learners/undocumented_bodies.rs

```diff
@@ -10,29 +10,69 @@ use crate::JsonTrail;
 
 #[derive(Default, Debug)]
 pub struct LearnedUndocumentedBodiesProjection {
-  observations_by_body_location: HashMap<BodyAnalysisLocation, TrailObservationsResult>,
+  request_response_observations: HashMap<BodyAnalysisLocation, TrailObservationsResult>,
+  query_params_observations: HashMap<BodyAnalysisLocation, TrailObservationsResult>,
 }
 
 impl LearnedUndocumentedBodiesProjection {
   fn with_body_analysis_result(&mut self, analysis: BodyAnalysisResult) {
-    let existing_observations = self
-      .observations_by_body_location
+    let collection = match analysis.body_location {
+      BodyAnalysisLocation::UnmatchedRequest { .. }
+      | BodyAnalysisLocation::UnmatchedResponse { .. } => &mut self.request_response_observations,
+      BodyAnalysisLocation::UnmatchedRequestQueryParameters { .. } => {
+        &mut self.query_params_observations
+      }
+      BodyAnalysisLocation::MatchedRequest { .. }
+      | BodyAnalysisLocation::MatchedResponse { .. } => {
+        unreachable!("analyzing undocumented bodies should only yield unmatched analysiss")
+      }
+    };
+
+    let existing_observations = collection
       .entry(analysis.body_location)
       .or_insert_with(|| TrailObservationsResult::default());
 
     existing_observations.union(analysis.trail_observations.normalized());
   }
 
   pub fn into_endpoint_bodies(
-    self,
+    mut self,
     id_generator: &mut impl SpecIdGenerator,
   ) -> impl Iterator<Item = EndpointBodies> {
     let mut endpoints_by_endpoint = HashMap::new();
-    for (body_location, observations) in self.observations_by_body_location {
+    for (body_location, observations) in self.request_response_observations {
       let (root_shape_id, body_commands) =
         observations.into_commands(id_generator, &JsonTrail::empty());
-      let endpoint_body =
-        EndpointBody::new(&body_location, root_shape_id, body_commands, id_generator);
+      let mut endpoint_body = EndpointBody::new(&body_location, root_shape_id, body_commands);
+
+      if let BodyAnalysisLocation::UnmatchedRequest {
+        ref path_id,
+        ref method,
+        ref content_type,
```

Comment:
What's does the usage of the `ref` keyword do here? And why is it necessary?

Reply:
It's to make sure `path_id`, `method` and `content_type` are captured as references to whatever is inside `body_location`, rather than moving them out (which is what would happen without, but I think you can also write without `ref` and instead match on `&body_location`). We want to do that, as `body_location` is used again later.